File: bintang/parser.py
# --- Kelas Error ---
class SyntaxError(Exception):
    """Kelas untuk error sintaksis yang lebih informatif."""
    def __init__(self, token, expected):
        # Asumsikan token memiliki posisi untuk info lebih lanjut
        pos = token.pos if hasattr(token, 'pos') else '?'
        message = f"Syntax Error: Unexpected token '{token.value}' ({token.type}) at pos {pos}. Expected: {expected}."
        super().__init__(message)

# --- Definisi Tipe Token (Lanjutan dari sebelumnya) ---
TT_INT    = 'INT'
TT_PLUS   = 'PLUS'
TT_MINUS  = 'MINUS'
TT_MUL    = 'MUL'
TT_DIV    = 'DIV'
TT_LPAREN = 'LPAREN'
TT_RPAREN = 'RPAREN'

# --- Kelas Token ---
class Token:
    def __init__(self, type_, value=None, pos=None):
        self.type = type_
        self.value = value
        self.pos = pos # Menambahkan posisi untuk error handling yang lebih baik

    def __repr__(self):
        if self.value:
            return f'{self.type}:{self.value}'
        return f'{self.type}'

# --- Kelas Lexer (Disesuaikan dengan penambahan posisi) ---
class Lexer:
    def __init__(self, text):
        self.text = text
        self.pos = -1
        self.current_char = None
        self.advance()

    def advance(self):
        self.pos += 1
        self.current_char = self.text[self.pos] if self.pos < len(self.text) else None

    def make_tokens(self):
        tokens = []

        while self.current_char != None:
            if self.current_char.isspace():
                self.advance()

            elif self.current_char.isdigit():
                tokens.append(self.make_number())

            elif self.current_char == '+':
                tokens.append(Token(TT_PLUS, value='+', pos=self.pos))
                self.advance()
            elif self.current_char == '-':
                tokens.append(Token(TT_MINUS, value='-', pos=self.pos))
                self.advance()
            elif self.current_char == '*':
                tokens.append(Token(TT_MUL, value='*', pos=self.pos))
                self.advance()
            elif self.current_char == '/':
                tokens.append(Token(TT_DIV, value='/', pos=self.pos))
                self.advance()
            elif self.current_char == '(':
                tokens.append(Token(TT_LPAREN, value='(', pos=self.pos))
                self.advance()
            elif self.current_char == ')':
                tokens.append(Token(TT_RPAREN, value=')', pos=self.pos))
                self.advance()

            else:
                raise Exception(f"Illegal Character: '{self.current_char}' at pos {self.pos}")

        # Tambahkan token akhir opsional (EOF) untuk mempermudah parser,
        # meskipun tidak wajib untuk parser validasi sederhana ini.
        return tokens

    def make_number(self):
        num_str = ''
        start_pos = self.pos # Simpan posisi awal angka

        while self.current_char != None and self.current_char.isdigit():
            num_str += self.current_char
            self.advance()

        return Token(TT_INT, int(num_str), pos=start_pos)


# --- Kelas Parser ---
class Parser:
    """
    Kelas yang bertanggung jawab membangun struktur pohon (AST) dan memvalidasi sintaks.
    Untuk saat ini hanya melakukan validasi sintaks.
    """
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = -1
        self.current_token = None
        self.advance()

    def advance(self):
        """Memajukan kursor ke token berikutnya."""
        self.pos += 1
        self.current_token = self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def parse(self):
        """Fungsi utama, memanggil aturan grammar tertinggi (expr)."""
        if self.current_token is None:
            # Input kosong dianggap valid (atau bisa kita buat tidak valid)
            return "Valid (Input Kosong)"

            # Panggil aturan grammar tertinggi (Expression)
        self.parse_expr()

        # Setelah parsing selesai, harusnya tidak ada token sisa (kecuali EOF)
        if self.current_token is None:
            return "Valid"
        else:
            # Jika masih ada token tersisa, berarti ada yang salah
            raise SyntaxError(self.current_token, "end of input")

    def parse_eat(self, token_type, expected_msg):
        """
        Helper untuk memastikan token saat ini sesuai harapan, lalu maju (eat)
        Jika tidak sesuai, lempar SyntaxError.
        """
        if self.current_token and self.current_token.type == token_type:
            self.advance()
        else:
            # Current token adalah token yang salah
            raise SyntaxError(self.current_token if self.current_token else Token("EOF", value="<End>", pos=self.pos), expected_msg)


    # --- Aturan Grammar (Recursive Descent) ---

    # factor : INT | LPAREN expr RPAREN
    def parse_factor(self):
        token = self.current_token

        if token and token.type == TT_INT:
            self.parse_eat(TT_INT, "INT")
            # Dalam parser AST sungguhan, kita akan mengembalikan node angka di sini

        elif token and token.type == TT_LPAREN:
            self.parse_eat(TT_LPAREN, "'('") # 'makan' (

            # Panggil aturan grammar yang lebih tinggi (Rekursif)
            self.parse_expr()

            self.parse_eat(TT_RPAREN, "')'") # harus diikuti oleh )

        else:
            # Jika bukan INT atau LPAREN, itu adalah Syntax Error
            raise SyntaxError(token if token else Token("EOF", value="<End>", pos=self.pos), "INT atau '('")


    # term : factor ((MUL | DIV) factor)*
    def parse_term(self):
        # 1. Selalu mulai dengan memanggil aturan dengan prioritas yang lebih tinggi (Factor)
        self.parse_factor()

        # 2. Loop untuk menangani operator * dan /
        while self.current_token and self.current_token.type in (TT_MUL, TT_DIV):
            op_token = self.current_token

            # 'Makan' operator * atau /
            self.parse_eat(op_token.type, f"'{op_token.value}'")

            # Harus diikuti oleh Factor lain (mempertahankan loop)
            self.parse_factor()


    # expr : term ((PLUS | MINUS) term)*
    def parse_expr(self):
        # 1. Selalu mulai dengan memanggil aturan dengan prioritas yang lebih tinggi (Term)
        self.parse_term()

        # 2. Loop untuk menangani operator + dan -
        while self.current_token and self.current_token.type in (TT_PLUS, TT_MINUS):
            op_token = self.current_token

            # 'Makan' operator + atau -
            self.parse_eat(op_token.type, f"'{op_token.value}'")

            # Harus diikuti oleh Term lain (mempertahankan loop)
            self.parse_term()


# --- Fungsi Utama untuk menjalankan Lexer dan Parser ---
def run_validation(text):
    print(f"\n--- Memproses Input: \"{text}\" ---")

    # Tahap 1: Lexical Analysis
    try:
        lexer = Lexer(text)
        tokens = lexer.make_tokens()
        print(f"Token: {tokens}")
    except Exception as e:
        print(f"❌ Lexer Error: {e}")
        return "Invalid"

    # Tahap 2: Parsing (Sintaks Analysis)
    try:
        parser = Parser(tokens)
        result = parser.parse()
        print(f"Result: {result}")
        return "Valid"
    except SyntaxError as e:
        print(f"❌ Parser Error: {e}")
        return "Invalid"


# --- Contoh Penggunaan dan Target Validasi ---

# Test Kasus 1: Valid
run_validation("10 + 2 * (5 - 3)")

# Test Kasus 2: Invalid (Syntax Error)
run_validation("10 + * ")

# Test Kasus 3: Invalid (Kurung tidak seimbang)
run_validation("(1 + 2")

# Test Kasus 4: Valid
run_validation("4 / 2 - 1")

File: irvan/parser.js
import { Lexer } from './lexer.js';

const TOKEN = {
    INT: "INT",
    PLUS: "PLUS",
    MINUS: "MINUS",
    MUL: "MUL",
    DIV: "DIV",
    LPAREN: "LPAREN",
    RPAREN: "RPAREN",
    IDENTIFIER: "IDENTIFIER",
    KEYWORD: "KEYWORD",
    EQ: "EQ",
    EOF: "EOF"
};

class NumberNode {
    constructor(value) {
        this.value = value;
    }

    toString() {
        return `${this.value}`;
    }
}

class BinOpNode {
    constructor(op, left, right) {
        this.op = op;
        this.left = left;
        this.right = right;
    }

    toString(indent = 0) {
        const pad = ' '.repeat(indent);
        return `BinOpNode(\n${pad}  ${this.left.toString(indent + 2)},\n${pad}  ${this.op},\n${pad}  ${this.right.toString(indent + 2)})`;
    }
}

class VarAccessNode {
    constructor(var_name_token) {
        this.var_name_token = var_name_token;
    }

    toString() {
        return `VarAccessNode(${this.var_name_token.value})`;
    }
}

class VarAssignNode {
    constructor(var_name_token, value_node) {
        this.var_name_token = var_name_token;
        this.value_node = value_node;
    }

    toString(indent = 0) {
        const pad = ' '.repeat(indent);
        return `VarAssignNode(${this.var_name_token.value}, ${this.value_node.toString(indent + 2)})`;
    }
}

export class Parser {
    constructor(input) {
        this.lexer = new Lexer(input);
        this.tokens = this.lexer.getTokens();
        this.tokens.push({ type: TOKEN.EOF, value: null });
        this.pos = 0;
        this.currentToken = this.tokens[this.pos];
    }

    error() {
        throw new Error(`Invalid`);
    }

    eat(tokenType) {
        if (this.currentToken.type === tokenType) {
            this.pos++;
            this.currentToken = this.tokens[this.pos];
        } else {
            this.error(`Diharapkan ${tokenType}, tapi mendapat ${this.currentToken.type}`);
        }
    }

    factor() {
        const token = this.currentToken;

        if (token.type === TOKEN.INT) {
            this.eat(TOKEN.INT);
            return new NumberNode(token.value);
        } else if (token.type === TOKEN.IDENTIFIER) {
            this.eat(TOKEN.IDENTIFIER);
            return new VarAccessNode(token);
        } else if (token.type === TOKEN.LPAREN) {
            this.eat(TOKEN.LPAREN);
            const result = this.expr();
            this.eat(TOKEN.RPAREN);
            return result;
        } else {
            this.error("Diharapkan integer, identifier, atau ekspresi dalam kurung");
        }
    }

    term() {
        let result = this.factor();

        while (this.currentToken.type === TOKEN.MUL || this.currentToken.type === TOKEN.DIV) {
            const token = this.currentToken;
            if (token.type === TOKEN.MUL) {
                this.eat(TOKEN.MUL);
                result = new BinOpNode(token.type, result, this.factor());
            } else if (token.type === TOKEN.DIV) {
                this.eat(TOKEN.DIV);
                result = new BinOpNode(token.type, result, this.factor());
            }
        }

        return result;
    }

    bin_op_expr() {
        let result = this.term();

        while (this.currentToken.type === TOKEN.PLUS || this.currentToken.type === TOKEN.MINUS) {
            const token = this.currentToken;
            if (token.type === TOKEN.PLUS) {
                this.eat(TOKEN.PLUS);
                result = new BinOpNode(token.type, result, this.term());
            } else if (token.type === TOKEN.MINUS) {
                this.eat(TOKEN.MINUS);
                result = new BinOpNode(token.type, result, this.term());
            }
        }

        return result;
    }

    expr() {
        if (this.currentToken.type === TOKEN.IDENTIFIER && this.tokens[this.pos + 1] && this.tokens[this.pos + 1].type === TOKEN.EQ) {
            const var_name = this.currentToken;
            this.eat(TOKEN.IDENTIFIER);
            this.eat(TOKEN.EQ);
            const value = this.expr();
            return new VarAssignNode(var_name, value);
        } else {
            return this.bin_op_expr();
        }
    }

    parse() {
        try {
            const result = this.expr();

            if (this.currentToken.type !== TOKEN.EOF) {
                this.error("Token tak terduga setelah ekspresi selesai");
            }

            return {
                valid: true,
                result: result,
                message: "Ekspresi valid"
            };
        } catch (error) {
            return {
                valid: false,
                result: null,
                message: error.message
            };
        }
    }

    static validate(input) {
        const parser = new Parser(input);
        return parser.parse();
    }
}
